pkg/tfvars/internal/cache/cache.go

```diff
@@ -0,0 +1,208 @@
+// <user_cache_dir>/openshift-installer/<dataType>_cache
+// If the directory doesn't exist, it will be automatically created.
+func getCacheDir(dataType string) (string, error) {
+	if dataType == "" {
+		return "", errors.Errorf("data type can't be an empty string")
+	}
+
+	userCacheDir, err := os.UserCacheDir()
+	if err != nil {
+		return "", err
+	}
+
+	cacheDir := filepath.Join(userCacheDir, applicationName, dataType+"_cache")
+
+	_, err = os.Stat(cacheDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			err = os.MkdirAll(cacheDir, 0755)
+			if err != nil {
+				return "", err
+			}
+		} else {
+			return "", err
+		}
+	}
+
+	return cacheDir, nil
+}
+
+// cacheKey returns a cache key based on file's etag
+func cacheKey(etag string) (key string, err error) {
+	if etag == "" {
+		return "", errors.Errorf("caching is not supported when ETag is unset")
+	}
+	etagSections := strings.SplitN(etag, "\"", 3)
+	if len(etagSections) != 3 {
+		return "", errors.Errorf("broken quoting: %s", etag)
+	}
+	if etagSections[0] == "W/" {
+		return "", errors.Errorf("caching is not supported for weak ETags: %s", etag)
+	}
+	opaque := etagSections[1]
+	if opaque == "" {
+		return "", errors.Errorf("caching is not supported when the opaque tag is unset: %s", etag)
+	}
+	hashed := md5.Sum([]byte(opaque))
+	return hex.EncodeToString(hashed[:]), nil
+}
+
+// cacheFile puts data in the cache
+func cacheFile(reader io.Reader, filePath string) (err error) {
+	logrus.Debugf("Unpacking file into %q...", filePath)
+
+	flockPath := fmt.Sprintf("%s.lock", filePath)
+	flock, err := os.Create(flockPath)
+	if err != nil {
+		return err
+	}
+	defer flock.Close()
+	defer func() {
+		err2 := os.Remove(flockPath)
+		if err == nil {
+			err = err2
+		}
+	}()
+
+	err = unix.Flock(int(flock.Fd()), unix.LOCK_EX)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err2 := unix.Flock(int(flock.Fd()), unix.LOCK_UN)
+		if err == nil {
+			err = err2
+		}
+	}()
+
+	_, err = os.Stat(filePath)
+	if err != nil && !os.IsNotExist(err) {
+		return nil // another cacheFile beat us to it
+	}
+
+	tempPath := fmt.Sprintf("%s.tmp", filePath)
+	file, err := os.OpenFile(tempPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0444)
+	if err != nil {
+		return err
+	}
+	closed := false
+	defer func() {
+		if !closed {
+			file.Close()
+		}
+	}()
+
+	_, err = io.Copy(file, reader)
+	if err != nil {
+		return err
+	}
+
+	err = file.Close()
+	if err != nil {
+		return err
+	}
+	closed = true
+
+	return os.Rename(tempPath, filePath)
+}
+
+// DownloadFile obtains a file from a given URL, puts it in the cache folder, defined by dataType parameter,
+// and returns the local file path.
+// If the query string contains sha256 parameter (i.e. https://example.com/data.bin?sha256=098a5a...),
+// then the downloaded data checksum will be compared with the provided value.
+func DownloadFile(baseURL string, dataType string) (string, error) {
+	// Send a request
+	resp, err := http.Get(baseURL)
```

Comment:
Is there a way to avoid the HTTP query at all? If we know the SHA, could we look to see if it's already in the cache before making any network connection out? Thinking of some specific scenarios with a  disconnected install. This looks like it downloads the whole image each time before populating the cache (i.e. the cache doesn't get used across multiple installer runs)?

Reply:
I added a check if a URL starts with "file://" we don't download anything and just reuse the existing file, like it happens now: https://github.com/openshift/installer/blob/master/pkg/tfvars/libvirt/cache.go#L32-L34

Also, we actually don't load any image data and immediately return the path if we found a file with ETag name in the cache. We just make a request and validate the response headers to find ETag there, that's all we do.

Reply:
The problem is any outbound query. For many bare metal environments we have a requirement that there can’t be any outbound internet connectivity at all. It’s possible to override the registry for containers, and for the image used by masters (which is different than bootstrap) we have a solution. The only one remaining is we don’t have a good solution that handles the qemu image for bootstrap. That is always an internet url in rhcos.json. Since we know the SHA and that’s definitive anyway, it makes sense to just see if we have a file named that already instead of using ETags at all. 

The idea being we populate the cache manually in a disconnected environment.

Reply:
Maybe my ideas and all that well thought out, I just saw this PR and thought it would be an easy change to just use the SHA all the time. Anyway, I don’t want to hijack your PR and I can always make a proposal separately to handle this case maybe by explicitly overriding the rhcos baseURI

Reply:
okay, I got you now... you're gonna laugh, but what you're saying was in my original patch! I calculated a filename corresponding to the url's md5 hash. If it helps you, I can bring that solution back.

Reply:
give me 10 minutes

Reply:
@stbenjam done! now you can calculate image filename in the cache by using `echo -n <image_url> | md5sum`

Reply:
awesome thanks so much!